README.md
# Chinatown Story
design rationale
We want to provide the users an opportunity to better know food industry of the Chinatown and its neighborhood near downtown Boston. At this point, we want to show some information that may be already available on Google or Yelp’s App, then add our own research and analysis of these dataset we are using (also combined with the Ssaki Chinatown project ）
1.	How did you choose your particular visual encodings, interaction, and animation techniques? 
We use Leaflet.js and Mapbox’s API to embody the coordinates onto a web-based map. We were to give a general impression of Chinatowns all over the world then move to our focus on Boston Area. The map can also be altered from 2D top view to a 3D perspective view, through slide down from upper pages to lower pages, with our re-designed template for visual style. We added filters like Price, Rating, Cuisine Code and Credit Score, to let user decide which markers to show up according to their purpose or demand. Lastly, we have an analytic chart/graph, created by d3.js, to show the discrete point that represent each company mainly based on their customers’ rating and the number of viewers(customers), but also include other information that related to the companies themselves.
2.	What alternatives did you consider and how did you arrive at your ultimate choices?

One design about the interactive map is that we were considering weather use popup windows to show the detailed information when the user’s mouse hover on a marker, while, we eventually choose to use the side panel to present these information by user clicking the marker. We believe in this way the layout is more user friendly and the panel would not just disappear if they want to point on the information in the panel 



sidebar/CHANGELOG.md
# Changelog

## v0.1.9 (2016-01-19)

- Fix unresolved variable error in `removeFrom()` method (see #38)


## v0.1.8 (2015-09-10)

- Fix text selection on sidebar (see #31)


## v0.1.7 (2015-09-09)

- Catch context menu events on sidebar (#23)
- Fix jQuery event handlers/Bootstrap controls on sidebar (#31)


## v0.1.6 (2014-08-23)

- Added support for [bower.io](http://bower.io/search/) (#21)


## v0.1.5 (2014-04-15)

- Added support for [component.io](http://component.io/) (#17, thanks @zedd45)
- Added `shown` and `hidden` events (#18)


## v0.1.4 (2014-02-16)

- Adjusted CSS for touch devices (#12)
- Converted build system to [gulp.js](http://gulpjs.com/)
- Added [TravisCI](http://travis-ci.org/) support for CI linting


## v0.1.3 (2014-02-06)

- Added `autoPan` option


## v0.1.2 (2013-12-28)

- Fixed broken `stopPropagation()` call (#2)
- Fire `show` and `hide` events on sidebar instances
- Added `getCloseButton()` method
- Several memory leaks related to event handlers fixed (#7)
  Thanks to Jerroyd Moore!


## v0.1.1 (2013-11-03)

- Fixed close button event handler (#1)


## v0.1.0 (2013-10-30)

- First official release
- `position` option
- `closeButton` option
- `getContainer()` and `setContent()` methods
- two more examples


## v0.0.0 (2013-10-29)

- Preview release


config.js
var config = {
    style: 'mapbox://styles/zoeyzhu/ck8d75c283p0j1itmy7uhufsn',
    accessToken: 'pk.eyJ1Ijoiem9leXpodSIsImEiOiJjazF2bGNzNmQwaWl1M2NxZ3kxNGJodDgxIn0.tDIH5cVr0SlDHxl6O4t4bA',
    showMarkers: false,
    theme: 'light',
    alignment: 'center',
    title: '',
    subtitle: '',
    byline:'' ,
    footer: '',
    chapters: [
        
        {
            id: 'ct_1',
            alignment:'Chinatown',
            title: 'Chinatown 1700',
            theme: 'black',
            size:'15px',
            alignment:'center',
            image: '',
            description: 'American Chinatown and its development derived from the flourishing of Chinese im- migrants bringing with their culinary and tra- ditional culture.</b>It is in the 1790s, when the first immigrants group emergent in the North America, that Chinese immigrants stepped on the stage of western migration history. ',
            location: {

                center: [1, 39.97790],
                zoom: 1.5,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: 'chinatown_1700',
                    opacity: 0.5
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinatown_1700',
                    opacity: 0
                }
            ]
        },
        {
            id: 'ct_2',
            alignment:'center',
            title: 'Chinatown 1800',
            theme: 'light',
            size:'15px',
            alignment:'center',
            image: '',
            description: 'The following 1850s witnessed the flooding Chinese labors from Pearl Delta River district heading for the California Golden Rush. In the 1860s, after the mania of gold going down, most of the labors found niches in the transcontinental railroad industry and agriculture. Not only has they brought supportive labor force for the nationwide mechanized transportation networks, which totally revolutionized the economic and demographic structure of west America, but also subverted the eating culture of Americans since they harnesses their fishery and agricultural experience to bring more ingredients onto American’s ta- ble. This phenomena directly contribute the to the emergence of Chinese group. With increasing number of Chinese immigrants and economic opportunities, Chinatown emerged.',
            location: {

                center: [1, 39.97790],
                zoom: 1.5,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: 'chinatown_1800',
                    opacity: 0.7
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinatown_1800',
                    opacity: 0
                }
            ]
        },

        {
            id: 'ct_3',
            alignment:'center',
            title: 'Chinatown 1900',
            theme: 'light',
            size:'15px',
            alignment:'center',
            image: '',
            description: 'Tracing back the story behind Chinatown, the first North American Chinese settlement was established in San Francisco, which served as a gateway for early Chinese immigrants from 1850s to 1900s. Initially, the purposes of Chinese immigrants , consisting of labors from Pearl River Delta, were the result of chances including California Gold Rush and Trans-continental Railroad . With the expansion and economic increasing, Chinatowns were founded in several big cities in North America including Los Angeles, New York, Vancouver and Boston. On the one hand, Chinatown has provided a harbor for Chinese Americans to support themself with help from neighborhood, on the other hand, Chinatown is where they can grab the food and merchandises from Far East.',
            location: {

                center: [1, 39.97790],
                zoom: 1.5,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0.5
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0
                }
            ]
        },


        {
            id: 'world',
            alignment:'center',
            title: 'Chinese Cuisine',
            theme: 'light',
            size:'15px',
            alignment:'center',
            image: 'pic/chinacuisine-02.png',
            description: 'There are 8 main regional dishes in China territory, ranging from speicy ;Chuang’to ‘Sour-Sweet’Yue’.Original Chinese dishes are preferred by  the first generations and food Neophyllics.',
            location: {

                center: [103.33331, 35.12050],
                zoom: 4,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: '',
                    opacity: 0.8
                }
            ],
            onChapterExit: [
                {
                    layer: 'phl-city-limits',
                    opacity: 0
                }
            ]
        },



     
        {
            id: 'China',
            title: 'The spreading of Chinese Cuisine',
            image: 'pic/cuisine-03.png',
            description: 'This steep, rocky gorge can be surprisingly technical. Follow the orange and yellow trails to repeatedly climb and descend through the schist hillsides (careful of the cliffs), or stick to the gravel Forbidden Drive for a relaxing ride along the creek. You\'ll forget you\'re in a city.',
            location: {
                center: [103.33331, 35.12050],
                zoom: 4,
                pitch: 0.00,
                bearing:0.00
            },
            onChapterEnter: [
                {
                    layer: 'wissahickon',
                    opacity: 1
                }
            ],
            onChapterExit: [
                {
                    layer: 'wissahickon',
                    opacity: 0
                }
            ]
        },

        {
            id: 'US',
            title: 'Chinese American Cuisine',
            image: 'pic/cuisine-04.png',
            showMarkers: true,
            description: 'Chinese-American Cuisine is built by American Chinese from 1924. These  new style of Chinese cuisine used readily available ingredients differing from traditional Chinese dishes. Frying, Choping and Grilling were the most common cooking techniques kept in A-C cookings.',
            location: {
                center: [-85, 40],
                zoom: 3.5,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0.7
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0
                }
            ]
        },

     

        {
            id: 'sf',
            title: 'Chinatown in San Fransisco',
            image: 'pic/immigration-01.png',
            showMarkers: true,
            description: 'The Chinatown in San Francisco is the oldest Chinatown in North America and the largest Chinese enclave outside Asia. It is also the oldest and largest of the four notable Chinese enclaves within San Francisco.Since its establishment in 1848, it has been highly important and influential in the history and culture of ethnic Chinese immigrants in North America. Chinatown is an enclave that continues to retain its own customs, languages, places of worship, social clubs, and identity. There are two hospitals, several parks and squares, numerous churches, a post office, and other infrastructure. ',
            location: {
                center: [-122.41639, 37.79370],
                zoom: 13.08,
                pitch: 47.50,
                bearing: 32.80
            },
            onChapterEnter: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0.7
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0
                }
            ]
        },

        {
            id: 'NYC',
            title: 'Chinatown in New York City',
            image: '',
            showMarkers: true,
            description: 'Manhattan"s Chinatown is a neighborhood in Lower Manhattan, New York City, bordering the Lower East Side to its east, Little Italy to its north, Civic Center to its south, and Tribeca to its west. With an estimated population of 90,000 to 100,000 people, Chinatown is home to the highest concentration of Chinese people in the Western Hemisphere. Manhattan"s Chinatown is also one of the oldest Chinese ethnic enclaves.The Manhattan Chinatown is one of nine Chinatown neighborhoods in New York City, as well as one of twelve in the New York metropolitan area, which contains the largest ethnic Chinese population outside of Asia, comprising an estimated 893,697 uniracial individuals as of 2017.',
            location: {
                center: [-73.99757, 40.71659],
                zoom: 13,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0.7
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinatown_1900',
                    opacity: 0
                }
            ]
        },
        
        {
            id: 'restaurant_03',
            title: 'Boston Chinatown',
            image: '',
            description: 'Chinatown, Boston is a neighborhood located in downtown Boston, Massachusetts. It is the only surviving historic ethnic Chinese enclave in New England since the demise of the Chinatowns in Providence, Rhode Island and Portland, Maine after the 1950s. Because of the high population of Asians and Asian Americans living in this area of Boston, there is an abundance of Chinese and Vietnamese restaurants located in Chinatown. It is one of the most densely populated residential areas in Boston and serves as the largest center of its East Asian and Southeast Asian cultural life. ',
            location: {
                center: [-71.05712, 42.35728],
                zoom: 14.08,
                pitch: 47.50,
                bearing: 32.80
            },
            onChapterEnter: [
                {
                    layer: 'chinese',
            
                    opacity: 0.7
                }
            ],
            onChapterExit: [
                {
                    layer: 'chinese',
                    opacity: 0
                }
            ]
        },
        
        {
            id: 'restaurant_03',
            title: 'Disppaering Chinatown Restaurant？',
            image: '',
            description: 'Filled with Chinese culture, including street vendor and people playing games under the Pagoda, today’s Chinatowns are still good places for both locals and visitors to experience Chinese food and culture in America. While there is a big controversy over decades due to the development of gentrification and the giant changing in real estate markets, the nostalgia stories behind Chinatowns have never gone away.',
            location: {
                center: [-71.05712, 42.35728],
                zoom: 14.08,
                pitch: 47.50,
                bearing: 32.80
            },
            onChapterEnter: [
                {
                    layer: 'cuisine',
            
                    opacity: 0.7
                }
            ],
            onChapterExit: [
                {
                    layer: 'cuisine',
                    opacity: 0
                }
            ]
        },

        {
            id: 'Check The Popular Restaurant',
            title: 'Check The Popular Restaurant',
            image: '',
            description: 'Therefore, several questions arouse, what is the next generation of Chinatown’s industry? what will happen in the city Boston with the gradual disappearance of Chinatown? Will the waste and transportation system be influenced with the gentrification of Chinatown ? What can government and community do to maintain the identity of Chinatown ? ',
            
            location: {
                center: [-71.05730, 42.35100],
                zoom: 15,
                pitch: 0.00,
                bearing: 0.00
            },
            onChapterEnter: [
                {
                    layer: 'restaurant',
                    opacity: 1
                }
            ],
            onChapterExit: [
                {
                    layer: 'restaurant',
                    opacity: 0.5
                }
            ]
        }
    ]
};


maparc.js




const map = new mapboxgl.Map({
        container: "map",
        style: "mapbox://styles/mapbox/light-v9",
        center: [-122.398, 37.788],
        zoom: 12,
        pitch: 60
    });

    map.on("style.load", () => {
        // Get data for the arc map from SFMTA origin/destination routes
        fetch('https://raw.githubusercontent.com/uber/deck.gl/master/examples/layer-browser/data/sfmta.routes.json')
            .then(function(response) {
                const arclayer = new MapboxLayer({
                    id: 'my-arc',
                    type: ArcLayer,
                    data: response.json(),
                    getSourcePosition: d => d.START,
                    getTargetPosition: d => d.END,
                    getSourceColor: d => [64, 255, 0],
                    getTargetColor: d => [0, 128, 200]
                });
          
                //Add the deck.gl arc layer to the map 
                map.addLayer(arclayer, 'waterway-label');
            });
    })

js/script.js
////////////////////////////////////////////////////////////
//////////////////////// Set-up ////////////////////////////
////////////////////////////////////////////////////////////

//Quick fix for resizing some things for mobile-ish viewers
var mobileScreen = ($( window ).innerWidth() < 500 ? true : false);

//Scatterplot
var margin = {left: 60, top: 20, right: 20, bottom: 60},
	width = Math.min($("#chart").width(), 840) - margin.left - margin.right,
	height = width*2/3;

var svg = d3.select("#chart").append("svg")
			.attr("width", (width + margin.left + margin.right))
			.attr("height", (height + margin.top + margin.bottom));
			
var wrapper = svg.append("g").attr("class", "chordWrapper")
			.attr("transform", "translate(" + margin.left + "," + margin.top + ")");

//////////////////////////////////////////////////////
///////////// Initialize Axes & Scales ///////////////
//////////////////////////////////////////////////////

var opacityCircles = 0.7,
	maxDistanceFromPoint = 50;

//Set the color for each region
var color = d3.scale.ordinal()
					.range(["#EFB605", "#E58903", "#E01A25", "#C20049", "#991C71", "#66489F", "#2074A0", "#10A66E", "#7EB852"])
					.domain(["American", "Asian", "Drink", "Chinese", 
							 "French", "Italian", "Japanese", "Seafood", "Other"]);
							 
//Set the new x axis range
var xScale = d3.scale.linear()
    .range([0, width])
    .domain([0, 2300]);
	
//Set new x-axis
var xAxis = d3.svg.axis()
	.orient("bottom")
	.ticks(2)
	.tickFormat(function (d) {
		return xScale.tickFormat((mobileScreen ? 4 : 8),function(d) { 
			var prefix = d3.formatPrefix(d); 
			return prefix.scale(d) + prefix.symbol;
		})(d);
	})	
	.scale(xScale);	
//Append the x-axis
wrapper.append("g")
	.attr("class", "x axis")
	.attr("transform", "translate(" + 0 + "," + height + ")")
	.call(xAxis);
		
//Set the new y axis range
var yScale = d3.scale.linear()
	.range([height,0])
	.domain(d3.extent(chinatown, function(d) { return d.rates; }))
	.nice();	
var yAxis = d3.svg.axis()
	.orient("left")
	.ticks(6)  //Set rough # of ticks
	.scale(yScale);	
//Append the y-axis
wrapper.append("g")
		.attr("class", "y axis")
		.attr("transform", "translate(" + 0 + "," + 0 + ")")
		.call(yAxis);
		
		
//Scale for the bubble size
var rScale = d3.scale.linear()
			.range([mobileScreen ? 1 : 2, mobileScreen ? 10 : 16])
			.domain(d3.extent(chinatown, function(d) { return d.price_num; }));

//////////////////////////////////////////////////////
///////////////// Initialize Labels //////////////////
//////////////////////////////////////////////////////

//Set up X axis label
wrapper.append("g")
	.append("text")
	.attr("class", "x title")
	.attr("text-anchor", "end")
	.style("font-size", (mobileScreen ? 8 : 12) + "px")
	.style('font-family',"Futura")
	.attr("transform", "translate(" + width + "," + (height - 10) + ")")
	.text("Reviews");

//Set up y axis label
wrapper.append("g")
	.append("text")
	.attr("class", "y title")
	.attr("text-anchor", "end")
	.style('font-family',"Futura")
	.style("font-size", (mobileScreen ? 8 : 12) + "px")
	.attr("transform", "translate(18, 0) rotate(-90)")
	.text("Rates of Restaurant(Google)");

////////////////////////////////////////////////////////////// 
//////////////////// Set-up voronoi ////////////////////////// 
////////////////////////////////////////////////////////////// 

//Initiate the voronoi function
//Use the same variables of the data in the .x and .y as used in the cx and cy of the circle call
//The clip extent will make the boundaries end nicely along the chart area instead of splitting up the entire SVG
//(if you do not do this it would mean that you already see a tooltip when your mouse is still in the axis area, which is confusing)
var voronoi = d3.geom.voronoi()
	.x(function(d) { return xScale(d.Reviews); })
	.y(function(d) { return yScale(d.rates); })
	.clipExtent([[0, 0], [width, height]]);

var voronoiCells = voronoi(chinatown);
	
////////////////////////////////////////////////////////////	
///////////// Circles to capture close mouse event /////////
////////////////////////////////////////////////////////////	

//Create wrapper for the voronoi clip paths
var clipWrapper = wrapper.append("defs")
    .attr("class", "clipWrapper");

clipWrapper.selectAll(".clip")
	.data(voronoiCells)
	.enter().append("clipPath")
  	.attr("class", "clip")
  	//.attr("id", function(d) { return "clip-" + d.point.company_code; })
  	.append("path")
  	.attr("class", "clip-path-circle")
  	//.attr("d", function(d) { return "M" + d.join(",") + "Z"; });

//Initiate a group element for the circles	
var circleClipGroup = wrapper.append("g")
	.attr("class", "circleClipWrapper"); 
	
//Place the larger circles to eventually capture the mouse
var circlesOuter = circleClipGroup.selectAll(".circle-wrapper")
	.data(chinatown.sort(function(a,b) { return b.price_num > a.price_num; }))
	.enter().append("circle")
	.attr("class", function(d,i) { return "circle-wrapper " + d.company_code; })
	.attr("clip-path", function(d) { return "url(#clip-" + d.company_code + ")"; })
    .style("clip-path", function(d) { return "url(#clip-" + d.company_code + ")"; })
	.attr("cx", function(d) {return xScale(d.Reviews);})
	.attr("cy", function(d) {return yScale(d.rates);})
	.attr("r", maxDistanceFromPoint)
	.on("mouseover", showTooltip)
	.on("mouseout",  removeTooltip);;

////////////////////////////////////////////////////////////	
/////////////////// Scatterplot Circles ////////////////////
////////////////////////////////////////////////////////////	

//Initiate a group element for the circles	
var circleGroup = wrapper.append("g")
	.attr("class", "circleWrapper"); 
	
//Place the chinatown circles
circleGroup.selectAll("chinatown")
	.data(chinatown.sort(function(a,b) { return b.price_num > a.price_num; })) //Sort so the biggest circles are below
	.enter().append("circle")
		.attr("class", function(d,i) { return "chinatown " + d.company_code; })
		.attr("cx", function(d) {return xScale(d.Reviews);})
		.attr("cy", function(d) {return yScale(d.rates);})
		.attr("r", function(d) {return rScale(d.price_num);})
		.style("opacity", opacityCircles)
		.style("fill", function(d) {return color(d.cuisine);});
			
///////////////////////////////////////////////////////////////////////////
///////////////////////// Create the Legend////////////////////////////////
///////////////////////////////////////////////////////////////////////////

if (!mobileScreen) {
	//Legend			
	var	legendMargin = {left: 5, top: 10, right: 5, bottom: 10},
		legendWidth = 145,
		legendHeight = 270;
		
	var svgLegend = d3.select("#legend").append("svg")
				.attr("width", (legendWidth + legendMargin.left + legendMargin.right))
				.attr("height", (legendHeight + legendMargin.top + legendMargin.bottom));			

	var legendWrapper = svgLegend.append("g").attr("class", "legendWrapper")
					.attr("transform", "translate(" + legendMargin.left + "," + legendMargin.top +")");
		
	var rectSize = 15, //dimensions of the colored square
		rowHeight = 20, //height of a row in the legend
		maxWidth = 144; //widht of each row
		  
	//Create container per rect/text pair  
	var legend = legendWrapper.selectAll('.legendSquare')  	
			  .data(color.range())                              
			  .enter().append('g')   
			  .attr('class', 'legendSquare') 
			  .attr("transform", function(d,i) { return "translate(" + 0 + "," + (i * rowHeight) + ")"; })
			  .style("cursor", "pointer")
			  .on("mouseover", selectLegend(0.02))
			  .on("mouseout", selectLegend(opacityCircles));
	 
	//Non visible white rectangle behind square and text for better hover
	legend.append('rect')                                     
		  .attr('width', maxWidth) 
		  .attr('height', rowHeight) 			  		  
		  .style('fill', "white")
		  .style('opacity',0);
	//Append small squares to Legend
	legend.append('rect')                                     
		  .attr('width', rectSize) 
		  .attr('height', rectSize) 			  		  
		  .style('fill', function(d) {return d;});                                 
	//Append text to Legend
	legend.append('text')                                     
		  .attr('transform', 'translate(' + 22 + ',' + (rectSize/2) + ')')
		  .attr("class", "legendText")
		  .style("font-size", "10px")
		  .style('font-family',"Futura")
		  .attr("dy", ".35em")		  
		  .text(function(d,i) { return color.domain()[i]; });  

	//Create g element for bubble size legend
	var bubbleSizeLegend = legendWrapper.append("g")
							.attr("transform", "translate(" + (legendWidth/2 - 30) + "," + (color.domain().length*rowHeight + 20) +")");
	//Draw the bubble size legend
	bubbleLegend(bubbleSizeLegend, rScale, legendSizes = [1,2,4], legendName = "Restaurant Price ($)");		
}//if !mobileScreen
else {
	d3.select("#legend").style("display","none");
}

//////////////////////////////////////////////////////
/////////////////// Bubble Legend ////////////////////
//////////////////////////////////////////////////////

function bubbleLegend(wrapperVar, scale, sizes, titleName) {

	var legendSize1 = sizes[0],
		legendSize2 = sizes[1],
		legendSize3 = sizes[2],
		legendCenter = 0,
		legendBottom = 60,
		legendLineLength = 25,
		textPadding = 5,
		numFormat = d3.format(",");
	
	wrapperVar.append("text")
		.attr("class","legendTitle")
		.attr("transform", "translate(" + legendCenter + "," + 0 + ")")
		.attr("x", 0 + "px")
		.style('font-family',"Futura")
		.attr("y", 0 + "px")
		.attr("dy", "1em")
		.text(titleName);
		
	wrapperVar.append("circle")
        .attr('r', scale(legendSize1))
        .attr('class',"legendCircle")
        .attr('cx', legendCenter)
        .attr('cy', (legendBottom-scale(legendSize1)));
    wrapperVar.append("circle")
        .attr('r', scale(legendSize2))
        .attr('class',"legendCircle")
        .attr('cx', legendCenter)
        .attr('cy', (legendBottom-scale(legendSize2)));
    wrapperVar.append("circle")
        .attr('r', scale(legendSize3))
        .attr('class',"legendCircle")
        .attr('cx', legendCenter)
        .attr('cy', (legendBottom-scale(legendSize3)));
		
	wrapperVar.append("line")
        .attr('class',"legendLine")
        .attr('x1', legendCenter)
        .attr('y1', (legendBottom-2*scale(legendSize1)))
		.attr('x2', (legendCenter + legendLineLength))
        .attr('y2', (legendBottom-2*scale(legendSize1)));	
	wrapperVar.append("line")
        .attr('class',"legendLine")
        .attr('x1', legendCenter)
        .attr('y1', (legendBottom-2*scale(legendSize2)))
		.attr('x2', (legendCenter + legendLineLength))
        .attr('y2', (legendBottom-2*scale(legendSize2)));
	wrapperVar.append("line")
        .attr('class',"legendLine")
        .attr('x1', legendCenter)
        .attr('y1', (legendBottom-2*scale(legendSize3)))
		.attr('x2', (legendCenter + legendLineLength))
        .attr('y2', (legendBottom-2*scale(legendSize3)));
		
	wrapperVar.append("text")
        .attr('class',"legendText")
        .attr('x', (legendCenter + legendLineLength + textPadding))
        .attr('y', (legendBottom-2*scale(legendSize1)))
		.attr('dy', '0.25em')
		.text("$");
	wrapperVar.append("text")
        .attr('class',"legendText")
        .attr('x', (legendCenter + legendLineLength + textPadding))
        .attr('y', (legendBottom-2*scale(legendSize2)))
		.attr('dy', '0.25em')
		.text("$$");
	wrapperVar.append("text")
        .attr('class',"legendText")
        .attr('x', (legendCenter + legendLineLength + textPadding))
        .attr('y', (legendBottom-2*scale(legendSize3)))
		.attr('dy', '0.25em')
		.text("$$$$");
		
}//bubbleLegend

///////////////////////////////////////////////////////////////////////////
//////////////////// Hover function for the legend ////////////////////////
///////////////////////////////////////////////////////////////////////////
	
//Decrease opacity of non selected circles when hovering in the legend	
function selectLegend(opacity) {
	return function(d, i) {
		var chosen = color.domain()[i];
			
		wrapper.selectAll(".chinatown")
			.filter(function(d) { return d.cuisine != chosen; })
			.transition()
			.style("opacity", opacity);
	  };
}//function selectLegend

///////////////////////////////////////////////////////////////////////////
/////////////////// Hover functions of the circles ////////////////////////
///////////////////////////////////////////////////////////////////////////

//Hide the tooltip when the mouse moves away
function removeTooltip (d, i) {

	//Save the chosen circle (so not the voronoi)
	var element = d3.selectAll(".chinatown."+d.company_code);
		
	//Fade out the bubble again
	element.style("opacity", opacityCircles);
	
	//Hide tooltip
	$('.popover').each(function() {
		$(this).remove();
	}); 
  
	//Fade out guide lines, then remove them
	d3.selectAll(".guide")
		.transition().duration(200)
		.style("opacity",  0)
		.remove();
		
}//function removeTooltip

//Show the tooltip on the hovered over slice
function showTooltip (d, i) {
	
	//Save the chosen circle (so not the voronoi)
	var element = d3.selectAll(".chinatown."+d.company_code);
	
	//Define and show the tooltip
	$(element).popover({
		placement: 'auto top',
		container: '#chart',
		trigger: 'manual',
		html : true,
		content: function() { 
			return "<span style='font-size: 15px; text-align: center; font-family:futura'>" +  d.Company_Name + "</span>" + "<br>" +
				   "<span style='font-size: 11px; text-align: center;'>" + "Address: " + d.Address + "</span>" + "<br>" +
				   "<span style='font-size: 11px; text-align: center;'>" + "Phone Number: " + d.Phone + "</span>" + "<br>" +
				   "<span style='font-size: 11px; text-align: center;'>" + "Website: " + d.Website + "</span>" + "<br>" +
				   "<span style='font-size: 11px; text-align: center;'>" + "Cuisine type: " + d.type + "</span>" + "<br>"; }
	});
	$(element).popover('show');

	//Make chosen circle more visible
	element.style("opacity", 1);

	//Place and show tooltip
	var x = +element.attr("cx"),
		y = +element.attr("cy"),
		color = element.style("fill");

	//Append lines to bubbles that will be used to show the precise data points
	
	//vertical line
	wrapper
		.append("line")
		.attr("class", "guide")
		.attr("x1", x)
		.attr("x2", x)
		.attr("y1", y)
		.attr("y2", height + 20)
		.style("stroke", color)
		.style("opacity",  0)
		.transition().duration(200)
		.style("opacity", 0.5);
	//Value on the axis
	wrapper
		.append("text")
		.attr("class", "guide")
		.attr("x", x)
		.attr("y", height + 38)
		.style("fill", color)
		.style("opacity",  0)
		.style("text-anchor", "middle")
		.text( d3.format(".2s")(d.Reviews) )
		.transition().duration(200)
		.style("opacity", 0.5);

	//horizontal line
	wrapper
		.append("line")
		.attr("class", "guide")
		.attr("x1", x)
		.attr("x2", -20)
		.attr("y1", y)
		.attr("y2", y)
		.style("stroke", color)
		.style("opacity",  0)
		.transition().duration(200)
		.style("opacity", 0.5);
	//Value on the axis
	wrapper
		.append("text")
		.attr("class", "guide")
		.attr("x", -25)
		.attr("y", y)
		.attr("dy", "0.35em")
		.style("fill", color)
		.style("opacity",  0)
		.style("text-anchor", "end")
		.text( d3.format(".1f")(d.rates) )
		.transition().duration(200)
		
		.style("opacity", 0.5);	
					
}//function showTooltip